### ISSUE TITLE
Sequence on variable/value

### ISSUE THREAD

**Issue by micabytes** [inkle/ink#57]
Ink has some pretty powerful mechanisms, but I'm missing one construct that I found very useful for writing texts with variables in my own scripting language. An example:
```
He brought me [# number | nothing! The cad|one calling bird|two french hens|three turtle doves|an aviary#].
```

This simple returns the first element in the construct if number is <= 0 (or false), the last element in the construct if number >= maximum elements, and the appropriate value if number is a valid index in the sequence.

E.g.,, for number = 0, you get:
He brought me nothing! The cad.
For number = 2, you get:
He brought me two french hens.
For numbers >= 4 you get:
He brought me an aviary.

You have something a little similar in your inline if-then `{ val : text1 | text0 }`, but that's of course not as powerful (`[# #] ` started out as a similar construct, before I decided to flip the order and generalize it), and I haven't noticed anything that provides the same functionality.

Perhaps I missed something? How do you handle that kind of variable-dependent writing in-line today?

**Comment by joethephish** [comment]
Interesting! So it's kinda an array indexer, except that it works beyond the boundaries too - that's pretty nice.

Nope, we don't support that right now, the closest thing we have is the general switch statement, which is multiline.

Could consider this though... @joningold?

**Comment by joningold** [comment]
That's a nice construct. We have no inline equivalent, but you can do

{
- val <= 0: nothing
- val = 1: a nut
...
- val >= n: pigs
}

And of course that can be extended to different and more specific
conditions if needed.

Still, there's something interesting about changing

{bool: true text | false text }

To

{int: zero text | one text | ... | nth text }

I guess the only thing I don't like is I wouldn't want to change the order
of the bool clauses to false | true.

In general... I don't find myself using this pattern much; since counters
are usually more woolly than exact. But that might just be me.

Cheers
Jon

**Comment by micabytes** [comment]
I use the scripts as the narrative side of a role-playing game engine, so there tend to be a lot of numbers involved. It's also very useful for stuff like:
```
He gave me [# $amount | nothing | one doubloon | $amount doubloons #].
```
Where $amount is the variable.

I ended up not having a separate boolean clause, since this one can do the same job, but I agree that it would be unfortunate to replace/overload the existing syntax.

How about simply adding it as a separate construct, like:
```
{# int_value : zero text | one text | ... | nth text }
```

That would seem to fit in nicely with the existing syntax.

**Comment by joningold** [comment]
We'll consider it. FWIW, I'd do the doubloon example as:


He gave me {amount == 0:nothing|{print_num(amount)} doubloon{amount > 1:s}}.


... where "print_num" is the function in the Writing with Ink doc that
turns an int to the appropriate string.

cheers
jon

**Comment by Fahrengeit** [comment]
It's a bit off topic, but still about sequences.
How do you manage situations when you need randomizing list of phrases, but you need to appear each only once.
It's for cases when you don't want the same order every time. For example, you are making a list of different descriptions or encounters (through the tunnel), you want them shuffled, but they shouldn't appear again later.

**Comment by michelmcbride** [comment]
_{# int_value : zero text | one text | ... | nth text }_ looks fine and logical but actually seems quite limiting to me? For example:

_He pressed the button for the [# elevator_button :  lobby | first | second | third | fourth | fifth | sixth | "oh crap I guess we can't reuse this dialogue for buildings with different numbers of floors" #] floor._
vs
_He pressed the button for the {elevator_button == 0 : lobby | {print_num_adjective(elevator_button)} floor}._

The latter seems much more elegant to me, when taking into account all the powerful directions one can take sentence structure when using ink. But maybe I'm just not understanding the more general use cases of the sequence idea?

**Comment by joningold** [comment]
Yeah, this one is tricky and I have a few patterns I use but none are
great. Most commonly I'd use diverts

- (top)
{
- random() < 0.1 && not a: -> a
- random() < 0.1 && not b: -> b
... More similar ...
}
-> top

Where random is a function you'd have to build for yourself (we may add one
at some point).

You could do something similar using "shuffle" lists.

I won't pretend that's compact or elegant! I usually reassure myself with
the idea that the only difference between a sequence and a shuffle is on
replay... But depends on what you're doing..!

Jon

**Comment by micabytes** [comment]
@michelmcbride It's basically an in-line switch on integers and there is nothing that requires the text for the different options to be the same (see the example in the first post). Sure, you could do the same with a function, but that is a suboptimal/heavyweight solution unless this is text that is going to be reused in many different places (and the same argument could be used against all the variable text constructs - they're all just convenient shorthand for something that could equally be done with functions).

As mentioned, my game engine is a role-playing game. Like many RPGs, that means that there are numbers everywhere - even actions/choices often result in numbers being generated (e.g., from skill rolls, difficulty checks, NPC rolls, etc). So I end up using this construct quite a lot.

Conversely, I have surprisingly little use for the existing sequence constructs, because of how my game is structured. As an example, in Pirates and Traders, the player can travel around in the Caribbean and visit islands/ports, where the land action is narrated. This means the player gets to see the same bits of narative text many times in descriptions of ports - which is alleviated by mixing up the text - but I rarely want to vary text based on how many times the player has seen a particular line of text (which is what the existing constructs facilitate), but rather dependent on either the "stats" of the island, or other variables. Even the shuffle is rarely useful in this context, because I typically want to shuffle on a random number rolled for that visit (which stays constant until the next visit), and not on the display of a particular line of text.

Different types of narrative uses, different needs.

I'm not dependent on you guys adding this to the core Ink code - that's one of the great benefits of OSS - so no pressure or rush to decide on that account. But assuming you guys like the construct, I would obviously prefer if it (or some equivalently useful alternative) would get introduced into "official" Ink eventually.

**Comment by michelmcbride** [comment]
Ah, I think I understand. Still, in terms of structure and readability I think I prefer the ink version of basically a bunch of specific and descriptive if/else statements rather than a switch that is limited to int. I'll leave it up to inkle!
